### ISSUE TITLE
darwin: http-parser doesn't build with sandbox enabled

### ISSUE THREAD

**Issue by datakurre** [NixOS/nixpkgs#10091]
I'm on master (https://github.com/NixOS/nixpkgs/commit/4a8e06f1f06fa0653793f038f5e2f53498ff7de5) on Darwin (OSX 10.10.5) with stdenvDarwinPure and building with chroot (Darwin sandbox). I seem to have issues with Python multiprocessing module (and probably with the sandbox). This error with building http-parser:

```
these derivations will be built:
  /nix/store/pzadiyj8wjmy9vkdnx8qhg7qp2gzw2y1-http-parser-2.5.0.drv
building path(s) ‘/nix/store/mvdhhjkk6rrvw4j2nml1x5ljv9m5fra7-http-parser-2.5.0’
unpacking sources
unpacking source archive /nix/store/pzrd7hdkhsyn8i0glnkchmp6k0n36gbn-v2.5.0.tar.gz
source root is http-parser-2.5.0
patching sources
applying patch /nix/store/x3rg29m48h5lkn200mysqz3kr8vq7rz8-build-shared.patch
patching file http_parser.gyp
configuring
sh: /usr/sbin/sysctl: Operation not permitted
Traceback (most recent call last):
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/bin/.gyp-wrapped", line 10, in <module>
    load_entry_point('gyp==0.1', 'console_scripts', 'gyp')()
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/__init__.py", line 538, in main
    return gyp_main(args)
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/__init__.py", line 514, in gyp_main
    options.duplicate_basename_check)
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/__init__.py", line 130, in Load
    params['parallel'], params['root_targets'])
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/input.py", line 2759, in Load
    check, generator_input_info)
  File "/nix/store/648d307hm1lm586dj380jdcvfq0cg9cv-python2.7-gyp-2015-06-11/lib/python2.7/site-packages/gyp/input.py", line 596, in LoadTargetBuildFilesParallel
    parallel_state.pool = multiprocessing.Pool(multiprocessing.cpu_count())
  File "/nix/store/a8hv9fl48qnx4nl8ji78xp2p856zz6pa-python-2.7.10/lib/python2.7/multiprocessing/__init__.py", line 136, in cpu_count
    raise NotImplementedError('cannot determine number of cpus')
NotImplementedError: cannot determine number of cpus
builder for ‘/nix/store/pzadiyj8wjmy9vkdnx8qhg7qp2gzw2y1-http-parser-2.5.0.drv’ failed with exit code 1
error: build of ‘/nix/store/pzadiyj8wjmy9vkdnx8qhg7qp2gzw2y1-http-parser-2.5.0.drv’ failed
```

And, on other occasions, seen seen this

```
  File "/nix/store/a8hv9fl48qnx4nl8ji78xp2p856zz6pa-python-2.7.10/lib/python2.7/multiprocessing/synchronize.py", line 59, in <module>
    " function, see issue 3770.")
ImportError: This platform lacks a functioning sem_open implementation, therefore, the required synchronization primitives needed will not function, see issue 3770.
```

**Comment by gridaphobe** [comment]
pure-darwin uses the OSX sandbox mechanism to restrict access to impure paths in the filesystem (among other things). It has a whitelist since there's some amount of impurity we'll always have to live with (eg libSystem).

I suspect `/usr/sbin/sysctl` is not on the whitelist by default, which would explain why the build is failing. IIRC the way to add paths to the whitelist is to use `__extraImpureHostDeps` attribute, e.g.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/pure-darwin/default.nix#L265

Can you try adding `__extraImpureHostDeps = [ "/usr/sbin/sysctl" ]` to the http-parser expression?

/cc @pikajude @copumpkin

**Comment by datakurre** [comment]
``__extraImpureHostDeps`` has no effect, but ``__impureHostDeps`` gives

```
building path(s) ‘/nix/store/vb5xg0im9nym08b15fab36dbcyn3ii2w-http-parser-2.5.0’
error: derivation '/nix/store/7lcgphrkmi6cmdcy9wg2imwqjnl9ivxs-http-parser-2.5.0.drv' requested impure path ‘/usr/sbin/sysctl’, but it was not in allowed-impure-host-deps (‘/System/Library /usr/lib /dev /bin/sh’)
```

**Comment by gridaphobe** [comment]
Ah, try adding `--option allowed-impure-host-deps ‘/System/Library /usr/lib /dev /bin/sh /usr/sbin/sysctl'` to your `nix-env` command. (we should really have an `extra-allowed-impure-host-deps` option..)

**Comment by datakurre** [comment]
It seems that I can also add that to ``/etc/nix/nix.conf``.

Now I get

```
  File "/nix/store/r496jzl2j8y5zwf0wpfnb4ych865cdss-python-2.7.10/lib/python2.7/multiprocessing/synchronize.py", line 75, in __init__
    sl = self._semlock = _multiprocessing.SemLock(kind, value, maxvalue)
OSError: [Errno 1] Operation not permitted
builder for ‘/nix/store/7lcgphrkmi6cmdcy9wg2imwqjnl9ivxs-http-parser-2.5.0.drv’ failed with exit code 1
```

And the following sandbox violation is logged:

```
28.9.2015 9.04.10,000 kernel[0]: Sandbox: python2.7(86945) deny ipc-posix-sem /mp86945-0
```

Is this no go with sandbox enabled?

**Comment by datakurre** [comment]
I like to continue to have sandbox on by default, but of course, here, I can just build http-parser without sandbox and go on.

Yet, @copumpkin might be interested about this (or know solution).

**Comment by datakurre** [comment]
@gridaphobe Thank you for showing that ``allowed-impure-host-deps`` is a configurable option.

**Comment by gridaphobe** [comment]
It may just mean that we need to make the sandbox script that we generate more configurable. IIRC much of it is hardcoded, and it's super restrictive. 

But Apple's sandbox mechanism is quite flexible. I think if you run nix in verbose (or maybe debug) mode, it will print the sandbox script.

**Comment by copumpkin** [comment]
I think there are two levels of solution here:
1. Compile `sysctl` ourselves: http://www.opensource.apple.com/source/system_cmds/system_cmds-643.30.1/sysctl.tproj/
2. Realize that most things that want `sysctl` are doing the sort of thing that we don't want them to do. In this case, it looks like `gyp` is trying to figure out how many CPUs we have so that presumably it can build things in parallel. We have a Nix variable for that, and `gyp` should be fixed to respect our variable.

Obviously, I'd prefer #2 :smile:

**Comment by copumpkin** [comment]
To clarify, we can add sandbox exceptions for "legitimate behavior" that happens during a regular build. But every thing we open up is another potential source of nondeterminism, so I'd urge us not to just start adding exceptions to fix errors, but rather to look at the root of the error to see why it wants to learn something about the system it's running on (since there's almost never a legitimate case for a build to know about its host)

**Comment by datakurre** [comment]
@copumpkin I tried packaging sysctl, but it appeared that the path is hardcoded in Python, your 2. option is probably the only reasonable way.

**Comment by pikajude** [comment]
@datakurre sounds like a job for `substituteInPlace`

**Comment by pikajude** [comment]
@copumpkin ipc-posix-sem is part of the POSIX IPC spec. I think it's probably okay to let packages do IPC (polyml does), and as you can see, it's not gyp doing the IPC stuff anyway. It looks like the multiprocessing library is, which doesn't surprise me.

Regardless, gyp is a giant ball of crap that we maybe need to write a whole stdenv preHook to handle.

**Comment by datakurre** [comment]
@pikajude @copumpkin Thanks. I managed to fix gyp. Will create pull fixing this issue.

**Comment by datakurre** [comment]
sysctl is here https://github.com/datakurre/nixpkgs/commit/0b075082939e25160b47784f4fc5f9853dec3268, but just fixing gyp was considered better

**Comment by datakurre** [comment]
Fixed in Nix 1.11 with fixes in darwin sandbox.
